### ISSUE TITLE
Feature Request: PPL Query Syntax support for Elasticsearch Plugin

### ISSUE THREAD

**Issue by eunice98k** [grafana/grafana#28674]
**What would you like to be added:**

This feature request proposes to extend Grafana’s Elasticsearch data source plugin with Piped Processing Language (PPL) support. [Pipe Processing Language](https://opendistro.github.io/for-elasticsearch-docs/docs/ppl/) is a semi-structured query language using a set of commands delimited by a vertical bar (|), and it can be used to query Elasticsearch when PPL plugin is installed for Elasticsearch. Adding PPL support will provide users with an alternative query syntax to Lucene, the only query syntax currently supported in the Elasticsearch data source plugin.
 
**Use Cases:**

The Elasticsearch plugin currently only supports Apache Lucene query syntax. PPL has an arguably more intuitive and flexible syntax than Lucene, and many users are already familiar with the [Unix pipe-style syntax](https://en.wikipedia.org/wiki/Pipeline_(Unix)). Particularly, PPL makes it easier to construct ad hoc queries, which would improve the user experience in Grafana Explore. 

Example of PPL query syntax:

```
source=accounts
| where age > 18
| fields firstname, lastname
```

**Proposal:**

PPL support could be optionally turned on as an Elasticsearch data source configuration option. When enabled, users will be able to choose between Lucene and PPL query syntax in the query editor. Different editor interfaces could be shown depending on the selected query syntax option. The end behavior would be the same for both query syntax choices—two functionally equivalent queries in different syntax should produce the same output.

Note that PPL queries are sent to a different endpoint exposed by the Elasticsearch plugin framework instead of the standard _search endpoint. Furthermore, the response structure is in [JDBC format](https://opendistro.github.io/for-elasticsearch-docs/docs/sql/protocol/#jdbc-format) and will need to be mapped to the required data frame structure. These would need to be addressed when implementing this feature.

More information regarding PPL query syntax and its commands can be found here: [PPL Reference Manual](https://github.com/opendistro-for-elasticsearch/sql/blob/develop/docs/experiment/ppl/index.rst)

cc @alolita @ahlaw

**Comment by alolita** [comment]
Hi Folks - We're getting ready to file our first PR on this feature request. Can you please assign this @ahlaw @eunice98k

**Comment by ahlaw** [comment]
Hi all, @eunice98k and I have been working on this feature and we would like to split it up into multiple PRs as the total changes are quite big. Would it be possible for someone authorized to create an intermediate remote branch like `elasticsearch-ppl-support` that we can target these PRs? This might make the review process easier as the split PRs are smaller. Once everything is merged into the intermediate branch, we can then have it reviewed to be merged into master.

**Comment by zoltanbedi** [comment]
I've created the branch for you. Please keep in mind that we are working on a big refactoring for elastic in https://github.com/grafana/grafana/pull/28033

**Comment by ahlaw** [comment]
Hmm that's super awkward as we have all the PRs already based on current master branch. Do you have a guesstimate on when that refactor will be merged?

**Comment by ahlaw** [comment]
Also, is the elasticsearch plugin still open to having features added before this refactor is finished?

**Comment by Elfo404** [comment]
@ahlaw We are currently in the process of reviewing the PR and we expect it to be merged to master no later than next week, therefore I would suggest using that as a base for PRs

**Comment by ahlaw** [comment]
Could someone with write permissions rebase the [branch](https://github.com/grafana/grafana/tree/elasticsearch-ppl-support) onto master so that PRs to it including the merged refactor don't become cluttered?

**Comment by Elfo404** [comment]
Closing this as it out of scope for the builtin ES plugin

**Comment by weaversam8** [comment]
If anyone tracking this issue was unaware, this feature exists in the [OpenSearch plugin for Grafana](https://grafana.com/grafana/plugins/grafana-opensearch-datasource/).
